internal/service/location/controller.go

```diff
@@ -85,7 +88,46 @@ func transformTollgates(tollgates []*tollgate.Tollgate) (result []*detector.Toll
 	return
 }
 
+func errorsToBadRequest(errors validation.Errors) *errdetails.BadRequest {
+	violations := make([]*errdetails.BadRequest_FieldViolation, 0)
+	for k, err := range errors {
+		violations = append(violations, &errdetails.BadRequest_FieldViolation{
+			Field:       k,
+			Description: err.Error(),
+		})
+	}
+	return &errdetails.BadRequest{
+		FieldViolations: violations,
+	}
+}
+
+type validator struct {
+	errors validation.Errors
+}
+
+func (v *validator) validate(name string, value interface{}, rules ...validation.Rule) {
+	err := validation.Validate(value, rules...)
+	if err != nil {
+		v.errors[name] = err
+	}
+}
+
 func (c *Controller) UpdateLocation(ctx context.Context, request *locationV1beta1.UpdateLocationRequest) (*locationV1beta1.UpdateLocationResponse, error) {
+	v := validator{errors: validation.Errors{}}
+	v.validate("worker_id", request.GetWorkerId(), validation.Required)
+	v.validate("timestamp", fmt.Sprintf("%v", request.GetTimestamp()), validation.Required, is.Int)
```

Comment:
Timestamp is a number, not a `google.protobuf.Timestamp`?

Reply:
yes, i will change that back.

 changed it while was playing with validations. wanted to have flat payload to use with client apps instead of nested struct 
{
  "update_time": {
    "seconds": 232432,
    "nanos": 10
  }
}

Reply:
You can use [GetSeconds](https://pkg.go.dev/google.golang.org/protobuf/types/known/timestamppb#Timestamp.GetSeconds):
```suggestion
	v.validate("timestamp", fmt.Sprintf("%v", request.GetTimestamp().GetSeconds()), validation.Required, is.Int)
```

Reply:
yes, exactly, as we go without structs, but only values validations, this is the way.